hometasks/2-building-blocks-of-oop-part-1/src/Triangle.ts

```diff
@@ -0,0 +1,58 @@
+import { Shape } from './Shape';
+import { Point } from './Point';
+
+export class Triangle extends Shape {
+    constructor(point1: Point, point2: Point, point3: Point);
+    constructor(point1: Point, point2: Point, point3: Point, color: string, filled: boolean);
+    constructor(point1: Point, point2: Point, point3: Point, color?: string, filled?: boolean) {
+        
+        if(!point1 || !point2 || !point3) {
+            throw new Error(`Points number is not correct`);
+        }
+
+        const points: Point[] = [];
+        points.push(point1);
+        points.push(point2);
+        points.push(point3);
+
+        if (color && filled) {
+            super(points, color, filled);
+        } else {
+            super(points);
+        }
+        
+    }
+
+    public toString(): string {
+        let objString = `Triangle[`
+        this.points.map( (point, index, arr) => {
+            objString = objString + `v${index+1}=(${point.getX}, ${point.getY})`;
+            if(index < arr.length - 1) {
+                objString += ',';
+            } else {
+                objString += ']';
+            }
+        });
+
+        return objString;
+    }
+
+    public getType(): string {
+        
+        const equalityLimit = 0.01;
+        const side1 = this.points[0].distance(this.points[1]);
+        const side2 = this.points[1].distance(this.points[2]);
+        const side3 = this.points[2].distance(this.points[0]);
```

Comment:
instead of code duplication, it's better to iterate through points and calculate the sides. And instead of checking each side if it's equal to another one, you can put all sides into Set, which will get rid of repeated things automatically. then you will be able check the size of the Set to define the type of triangle

You can use reducer in order to calculate unique sides by only one iteration:


```
const uniqueSides = this.points.reduce((acc, point, index)=> {
   const next = (index + 1) % this.points.length;
   const nextPoint = this.points[next];

   const side = point.distance(nextPoint);
   acc.add(side);
}, new Set());

if(uniqueSides.size === 3) return ...
```